rs/execution_environment/src/canister_manager.rs

```diff
@@ -1525,67 +1489,42 @@ impl CanisterManager {
             .memory_usage()
             .saturating_add(&new_snapshot_size)
             .saturating_sub(&replace_snapshot_size);
-        let validated_memory_usage = match self.memory_usage_checks(
+        let validated_memory_usage = self.memory_usage_checks(
             subnet_size,
             canister,
             round_limits,
             new_memory_usage,
             old_memory_usage,
             resource_saturation,
-        ) {
-            Ok(validated_memory_usage) => validated_memory_usage,
-            Err(err) => {
-                return (Err(err), NumInstructions::from(0));
-            }
-        };
+        )?;
 
         // Charge for taking a snapshot of the canister.
         let instructions = self
             .config
             .canister_snapshot_baseline_instructions
             .saturating_add(&new_snapshot_size.get().into());
 
-        if let Err(err) = self.cycles_account_manager.consume_cycles_for_instructions(
-            &sender,
-            canister,
-            instructions,
-            subnet_size,
-            // For the `take_canister_snapshot` operation, it does not matter if this is a Wasm64 or Wasm32 module
-            // since the number of instructions charged depends on constant set fee and snapshot size
-            // and Wasm64 does not bring any additional overhead for this operation.
-            // The only overhead is during execution time.
-            WasmExecutionMode::Wasm32,
-        ) {
-            return (
-                Err(CanisterManagerError::CanisterSnapshotNotEnoughCycles(err)),
-                0.into(),
-            );
-        };
+        // TODO [EXC-2046]: Split into 'check' and 'perform'
+        self.cycles_account_manager
+            .consume_cycles_for_instructions(
+                &sender,
+                canister,
+                instructions,
+                subnet_size,
+                // For the `take_canister_snapshot` operation, it does not matter if this is a Wasm64 or Wasm32 module
+                // since the number of instructions charged depends on constant set fee and snapshot size
+                // and Wasm64 does not bring any additional overhead for this operation.
+                // The only overhead is during execution time.
+                WasmExecutionMode::Wasm32,
+            )
+            .map_err(CanisterManagerError::CanisterSnapshotNotEnoughCycles)?;
 
         // Create new snapshot.
-        let new_snapshot = match CanisterSnapshot::from_canister(canister, state.time())
-            .map_err(CanisterManagerError::from)
-        {
-            Ok(s) => s,
-            Err(err) => return (Err(err), instructions),
-        };
+        let new_snapshot = CanisterSnapshot::from_canister(canister, state.time())
```

Comment:
Could you please leave a comment at `CanisterSnapshot::from_canister` that it must be cheap upon error?